### ISSUE TITLE
New error (sadly)

### ISSUE THREAD

**Issue by holgafreak** [torch/nn#981]
got further, but noiw this one:

In 1 module of nn.Sequential:
/home/xxx/torch/install/share/lua/5.1/nn/THNN.lua:110: bad argument #3 to 'v' (cannot convert 'struct THCudaTensor *' to 'struct THFloatTensor *')
stack traceback:
	[C]: in function 'v'
	/home/xxx/torch/install/share/lua/5.1/nn/THNN.lua:110: in function 'SpatialConvolutionMM_updateOutput'
	...in/torch/install/share/lua/5.1/nn/SpatialConvolution.lua:96: in function <...in/torch/install/share/lua/5.1/nn/SpatialConvolution.lua:92>

yesterday didn't have this error either

-m

**Comment by soumith** [comment]
have you made sure that your model is typecasted to CUDA (or that your input is Float?)

**Comment by holgafreak** [comment]
yes it is

**Comment by supakjk** [comment]
Same problem for ClassNLLCriterion after the update. (It worked fine until yesterday.)
luajit: /xxxxx/torch/install/share/lua/5.1/nn/THNN.lua:110: bad argument #3 to 'v' (cannot convert 'struct THCudaTensor *' to 'struct THCudaLongTensor *')
stack traceback:
        [C]: in function 'v'
        /xxxxx/torch/install/share/lua/5.1/nn/THNN.lua:110: in function 'ClassNLLCriterion_updateOutput'
        .../torch/install/share/lua/5.1/nn/ClassNLLCriterion.lua:41: in function 'updateOutput'
        ...torch/install/share/lua/5.1/nn/CrossEntropyCriterion.lua:13: in function 'forward'
        xxxxx.lua:99: in main chunk
        [C]: at 0x00405bb0

**Comment by soumith** [comment]
@supakjk update nn and cunn both.

**Comment by soumith** [comment]
@holgafreak can you give a small test case for this?

**Comment by holgafreak** [comment]
@soumith after updating nn and cunn everything goes ok with my code.
 But for some reason the labels go wild, and i'm getting this assertion cur_target > 0 and cur_target <= n_classes  in v-function. (I'm just remembering something like this type of error, it's on the linux-thing). Labels on my code should just be  1 and 2, and they are not changed anywhere. Printing them the values are ints, but  mostly >10000.
tried on os x and here's one from the demos train-on-cifar:
qlua: /Users/mjkoskin/torch-cl/install/share/lua/5.1/nn/THNN.lua:807: Assertion `cur_target >= 0 && cur_target < n_classes' failed.  at /Users/mjkoskin/torch-cl/extra/nn/lib/THNN/generic/ClassNLLCriterion.c:31

**Comment by supakjk** [comment]
Even after updating nn and cunn, mine still doesn't work.

`require 'nn'
torch.setdefaulttensortype('torch.FloatTensor')
crit = nn.CrossEntropyCriterion()
input = torch.Tensor():rand(5)
label = 3
crit:forward(input, label)
`
works without problem but the following code (CUDA version) produces the error message I mentioned above.
`require 'cunn'
torch.setdefaulttensortype('torch.FloatTensor')
crit = nn.CrossEntropyCriterion():cuda()
input = torch.CudaTensor():randn(5)
label = 3
crit:forward(input, label)
`

**Comment by 1byxero** [comment]
I was trying [this](https://github.com/soumith/cvpr2015/blob/master/Deep%20Learning%20with%20Torch.ipynb) tutorial of torch and was trying to execute CNN for cifar10 given on the page and I encountered a similar error... Please help
`/home/cuda/torch/install/share/lua/5.1/nn/THNN.lua:110: bad argument #3 to 'v' (cannot convert 'struct THCudaTensor *' to 'struct THCudaLongTensor *')

stack traceback:
	[C]: in function 'v'
	/home/cuda/torch/install/share/lua/5.1/nn/THNN.lua:110: in function 'ClassNLLCriterion_updateOutput'
	...uda/torch/install/share/lua/5.1/nn/ClassNLLCriterion.lua:41: in function 'forward'
	...da/torch/install/share/lua/5.1/nn/StochasticGradient.lua:35: in function 'train'
	[string "_RESULT={trainer:train(trainset)}"]:1: in main chunk
	[C]: in function 'xpcall'
	/home/cuda/torch/install/share/lua/5.1/trepl/init.lua:652: in function 'repl'
	...cuda/torch/install/lib/luarocks/rocks/trepl/scm-1/bin/th:199: in main chunk
	[C]: at 0x00406670	`

**Comment by supakjk** [comment]
The problem seems because of the target variable is not correctly set to be a CudaLongTensor.
As a temporary solution, I did something like the following. Hope there would be certain update or comments regarding this issue.
local theCrit = nn.CrossEntropyCriterion():cuda()
theCrit.nll.target = torch.CudaLongTensor{theCrit.nll.target[1]}

**Comment by GenTxt** [comment]
Hi supakjk:

I have the same error and torch/lua is new to me.  Could you explain where I add the code for your temporary solution? 

Is it in the module(s) listed in the error or the lua script I'm trying to run, or both?

An example would be appreciated.

Thanks

Error:

/home/gentxt/torch/install/share/lua/5.1/nn/THNN.lua:110: bad argument #3 to 'v' (cannot convert 'struct THCudaTensor *' to 'struct THCudaLongTensor *')
stack traceback:
	[C]: in function 'v'
	/home/gentxt/torch/install/share/lua/5.1/nn/THNN.lua:110: in function 'ClassNLLCriterion_updateOutput'
	...txt/torch/install/share/lua/5.1/nn/ClassNLLCriterion.lua:41: in function 'forward'
	sample.lua:182: in function 'sample'
	sample.lua:236: in main chunk
	[C]: in function 'dofile'
	...usr/torch/install/lib/luarocks/rocks/trepl/scm-1/bin/th:145: in main chunk
	[C]: at 0x00405e40

Torch/lua works fine with torch-rnn, https://github.com/karpathy/char-rnn etc. but doesn't work with the current code I'm trying to run. I've updated everything and the error continues.

**Comment by supakjk** [comment]
I mean, when you get an instance of the cuda version of ClassNLLCriterion, change the target field of that instance to be a CudaLongTensor  (probably initially a CudaTensor)
I think it should be automatically done but the current source code missed that part.

**Comment by soumith** [comment]
it is automatically done in the source code: https://github.com/torch/nn/blob/master/ClassNLLCriterion.lua#L36

**Comment by soumith** [comment]
actually, i realized that i missed the non-batch case. I'm fixing it now.

**Comment by soumith** [comment]
this should be fixed now in master, and reinstalling the "nn" package will make it go away.

https://github.com/torch/nn/commit/5ade793bf2ade45774d38fe6bfe7368c9fddf567

luarocks install nn

**Comment by dkbemisIII** [comment]
This still breaks for me, after updating. The code posted by @supakjk above gives the same error. torch.cudaLong is nil for me. Do you mean:

 self.target = target.cudaLong and self.target:cudaLong() or self.target:cuda()

**Comment by supakjk** [comment]
It should be torch.CudaLong (not torch.cudaLong.)

**Comment by dkbemisIII** [comment]
Don't think so. target.cudaLong checks to make sure the conversion function is present for the tensor. Possibly 'torch.CudaLongTensor' could serve the same purpose, but indirectly. torch.CudaLong is still nil.

**Comment by supakjk** [comment]
My mistake. I mean torch.CudaLongTensor.
After updating all the related ones (nn,cunn,torch,cutorch,etc), my test code above worked fine.

**Comment by dkbemisIII** [comment]
That's a little surprising. There was a first fix that should have worked, but broke compatibility when there was no CudaLong. The next update seems to have rebroken the initial issue (at least for me), because of the typo.

If you update nn, it works for you?

**Comment by soumith** [comment]
there was my patch which was the attempted fix. then @mys007 sent a fix for back-compat, but it was broken. i then pushed another fix on top of his fix that keeps back-compat and works for master too.

**Comment by soumith** [comment]
if you now update nn, it should (fingers-crossed) work for you

**Comment by dkbemisIII** [comment]
Seems good now. Thanks.

**Comment by uahsan3** [comment]
When I rebuild cunn, I get the following:

[ 14%] Building NVCC (Device) object lib/THCUNN/CMakeFiles/THCUNN.dir/THCUNN_generated_SpatialDilatedConvolution.cu.o
Building NVCC (Device) object lib/THCUNN/CMakeFiles/THCUNN.dir/THCUNN_generated_AbsCriterion.cu.o
/tmp/luarocks_cunn-scm-1-258/cunn/lib/THCUNN/RReLU.cu(68): error: identifier "THCRandom_generatorStates" is undefined

1 error detected in the compilation of "/tmp/tmpxft_00003e0f_00000000-7_RReLU.cpp1.ii".
CMake Error at THCUNN_generated_RReLU.cu.o.cmake:267 (message):
  Error generating file
  /tmp/luarocks_cunn-scm-1-258/cunn/build/lib/THCUNN/CMakeFiles/THCUNN.dir//./THCUNN_generated_RReLU.cu.o


make[2]: *** [lib/THCUNN/CMakeFiles/THCUNN.dir/THCUNN_generated_RReLU.cu.o] Error 1
make[2]: *** Waiting for unfinished jobs....
make[1]: *** [lib/THCUNN/CMakeFiles/THCUNN.dir/all] Error 2
make: *** [all] Error 2

Error: Build error: Failed building.

Any suggestions?

**Comment by soumith** [comment]
@uahsan3 this comes because of an outdated cutorch version.
luarocks install cutorch
luarocks install cunn

**Comment by yuvalpinter** [comment]
@soumith thanks!

**Comment by eriche2016** [comment]
Hi, I use nn.LookupTable module , but came accross the similiar error above, can anyone fix this?

**Comment by eriche2016** [comment]
my erorr message is below:
ome/xxx/torch/install/bin/luajit: /home/xxx/.luarocks/share/lua/5.1/nn/THNN.lua:109: bad argument #2 to 'v' (cannot convert 'struct THCudaTensor *' to 'struct THCudaLongTensor *')
stack traceback:
        [C]: in function 'v'
        /home/xxx/.luarocks/share/lua/5.1/nn/THNN.lua:109: in function 'LookupTable_accGradParameters'
        /home/xxx/.luarocks/share/lua/5.1/nn/LookupTable.lua:85: in function 'accGradParameters'
        /home/xxx/.luarocks/share/lua/5.1/nn/Module.lua:32: in function 'backward'
        ./misc_saver2_reg_atten_ws/LanguageModel.lua:738: in function 'updateGradInput'
        /home/xxx/.luarocks/share/lua/5.1/nn/Module.lua:31: in function 'backward'
        train_reg_on_att.lua:496: in function 'lossFun'
        train_reg_on_att.lua:574: in main chunk
        [C]: in function 'dofile'
        .../hxw/torch/install/lib/luarocks/rocks/trepl/scm-1/bin/th:145: in main chunk
        [C]: at 0x00406670

**Comment by eriche2016** [comment]
an i have update my torch, nn, cutorch, cunn to the latest versions. any idea?

**Comment by fmassa** [comment]
@eriche2016 pass a `CudaLongTensor` instead of a `CudaTensor`, and you should be fine.

**Comment by eriche2016** [comment]
Still, i got the error. Below is the test code, check it. 
th> model = nn.LookupTable(4, 3)
                                                                      [0.0001s]
th> model:cuda()
nn.LookupTable
                                                                      [0.0031s]
th> model:forward(torch.CudaLongTensor({1}))
 0.3487  1.1548 -0.7722
[torch.CudaTensor of size 1x3]

                                                                      [0.0016s]
th> model:backward(torch.CudaLongTensor{1}, torch.CudaTensor(1, 4))
/home/hxw/.luarocks/share/lua/5.1/nn/THNN.lua:109: bad argument #2 to 'v' (cannot convert 'struct THCudaTensor *' to 'struct THCudaLongTensor *')
stack traceback:
        [C]: in function 'v'
        /home/hxw/.luarocks/share/lua/5.1/nn/THNN.lua:109: in function 'LookupTable_accGradParameters'
        /home/hxw/.luarocks/share/lua/5.1/nn/LookupTable.lua:85: in function 'accGradParameters'
        /home/hxw/.luarocks/share/lua/5.1/nn/Module.lua:32: in function 'backward'
        [string "_RESULT={model:backward(torch.CudaLongTensor{..."]:1: in main chunk
        [C]: in function 'xpcall'
        /home/hxw/torch/install/share/lua/5.1/trepl/init.lua:650: in function 'repl'
        .../hxw/torch/install/lib/luarocks/rocks/trepl/scm-1/bin/th:199: in main chunk
        [C]: at 0x00406670

**Comment by fmassa** [comment]
@eriche2016 It seems that you don't have the latest `nn`. Line 85 in `LookupTable.lua` [does not match yours](https://github.com/torch/nn/blob/master/LookupTable.lua#L85).

**Comment by eriche2016** [comment]
@fmassa  I update my nn with command below:
```luarocks install nn```
I suppose it is the latest.

**Comment by fmassa** [comment]
@eriche2016 Then I don't understand your error message. The `distro` package points latest `nn` to [this line](https://github.com/torch/nn/blob/a8e63f2da3d3d84a7e1eed917572901a9ffba5d9/LookupTable.lua#L85), which doesn't correspond to the error you are facing.
I'd check if the installation was successful, or if you had errors during compilation.

**Comment by eriche2016** [comment]
I open my LookupTable file, and it is the latest. see blow
![image](https://cloud.githubusercontent.com/assets/11784910/19560654/e5ae6494-9707-11e6-9c8c-0ebfc6a83588.png)

**Comment by fmassa** [comment]
@eriche2016 there seems to be something wrong with your setup. The error message that you show corresponds to a line of comment.

**Comment by eriche2016** [comment]
Oh, i got it, the error its sit on the file in:
/home/xxx/.luarocks/share/lua/5.1/nn/LookupTable.lua:85
not in /home/xxx/torch/ folder, which contains the latest nn.  I will have to fix this

**Comment by eriche2016** [comment]
@fmassa  thank u very much for your patience. problem solved
